### ISSUE TITLE
get 0 gene symbols match between the datasets and the BLAST graph

### ISSUE THREAD

**Issue by huawen-poppy** [atarashansky/SAMap#59]
Hello.

I followed the tutorial to run the SAMap. I tried to compare the given data hydra and my data. However, after running the first step (load and preprocess the data), it turns out that the algorithm finds 0 gene symbols match between the datasets and the BLAST graph of my data while it finds 15363 `hy` gene symbols match between the datasets and the BLAST graph of the hydra.
My fasta file headers are the same as the gene names in my h5ad file. How can I fix this problem?

```
fn1 = "chr_data/hydra.h5ad"

fn2 = 'chr_data/cc7_raw_Samap_ann.h5ad'

filenames = {'hy':fn1,'AI':fn2}

sm = SAMAP(
        filenames,
        f_maps = 'chr_data/maps/',
        save_processed=True 
    )
```

> Processing data hy from:
> chr_data/hydra.h5ad
> RUNNING SAM
> Iteration: 0, Convergence: 1.0
> Iteration: 1, Convergence: 0.8913929123492299
> Iteration: 2, Convergence: 0.03301924361455135
> Computing the UMAP embedding...
> Elapsed time: 11.067883014678955 seconds
> Not updating the manifold...
> Processing data AI from:
> chr_data/cc7_raw_Samap_ann.h5ad
> RUNNING SAM
> Iteration: 0, Convergence: 1.0
> Iteration: 1, Convergence: 0.8968904643820371
> Iteration: 2, Convergence: 0.012796760284338031
> Computing the UMAP embedding...
> Elapsed time: 46.4692497253418 seconds
> Not updating the manifold...
> 15363 `hy` gene symbols match between the datasets and the BLAST graph.
> 0 `AI` gene symbols match between the datasets and the BLAST graph.

**Comment by atarashansky** [comment]
Sorry you’re experiencing trouble!

could you output this after creating the SAMAP object?

sm.sams[‘AI’].adata.var_names

And could you go into the ‘maps’ directory you created with BLAST and output the first few lines of the txt files created in the hyAI (or AIhy) folder?

**Comment by huawen-poppy** [comment]
Yes, in the SAMAP object its gene name somehow changed. 
`sm.sams['AI'].adata.var_names`

> Index(['AI_0', 'AI_1', 'AI_2', 'AI_3', 'AI_4', 'AI_5', 'AI_6', 'AI_7', 'AI_8',
       'AI_9',
       ...
       'AI_29259', 'AI_29260', 'AI_29261', 'AI_29262', 'AI_29263', 'AI_29264',
       'AI_29265', 'AI_29266', 'AI_29267', 'AI_29268'],
      dtype='object', length=29269)

`cc7=sc.read_h5ad("chr_data/cc7_raw_Samap_ann.h5ad")
cc7.var_names`

> Index(['AIPGENE5422', 'AIPGENE22785', 'AIPGENE22786', 'AIPGENE19183',
       'AIPGENE12227', 'AIPGENE12226', 'AIPGENE12228', 'AIPGENE19834',
       'AIPGENE19833', 'AIPGENE19832',
       ...
       'AIPGENE4037', 'AIPGENE3960', 'AIPGENE3980', 'AIPGENE4044',
       'AIPGENE4024', 'AIPGENE3946', 'AIPGENE4087', 'AIPGENE3964',
       'AIPGENE3936', 'AIPGENE4012'],
      dtype='object', length=29269)

`B = pd.read_csv("./chr_data/maps/hyAI/AI_to_hy.txt",sep='\t',index_col=0,header=None)
B.head()`

> ![image](https://user-images.githubusercontent.com/70507602/143482063-4ed8471f-106c-4a42-8f84-fed752ec9f43.png)

**Comment by atarashansky** [comment]
Can you try this:

from samalg import SAM
sam = SAM()
sam.load_data(fn2)
sam.adata.var_names

**Comment by huawen-poppy** [comment]
Yesss, below is the output:

> ![image](https://user-images.githubusercontent.com/70507602/143499403-b5f91eb9-0603-44f2-918d-2f29706f993c.png)

**Comment by atarashansky** [comment]
Can you try loading the processed data?

It should be saved under `chr_data/cc7_raw_Samap_ann_pr.h5ad`. Are the var_names correct for that one?

**Comment by huawen-poppy** [comment]
The var_names in the processed data are not correct.

> 
![image](https://user-images.githubusercontent.com/70507602/144101324-3e9f4a39-ca17-4959-8362-3a3d60e69451.png)

**Comment by atarashansky** [comment]
Can you try this? I'm still having trouble reproducing the issue on my end.
```
from samalg import SAM
sam = SAM()
sam.load_data(fn2)
sam.preprocess_data(
    sum_norm="cell_median",
    norm="log",
    thresh_low=0.0,
    thresh_high=0.96,
    min_expression=1,
)
sam.run(
    preprocessing="StandardScaler",
    npcs=100,
    weight_PCs=False,
    k=20,
    n_genes=3000,
    weight_mode='rms'
)
print(sam.adata.var_names)
```

**Comment by huawen-poppy** [comment]
Yesss, after running the code, its output shows like below:

> 
![image](https://user-images.githubusercontent.com/70507602/145016670-2554dcb3-e05f-438f-bf50-befdc6ed6fc8.png)

**Comment by atarashansky** [comment]
Hmm, I am trying the above snippet on my datasets and I'm not able to reproduce the issue.

Can you check which versions of SAM and anndata you have? 

```
import samalg
import anndata
print(anndata.__version__)
print(samalg.__version__)
```

**Comment by huawen-poppy** [comment]
yess, my anndata's version is 0.7.5 and my samalg's version is 0.8.9
![image](https://user-images.githubusercontent.com/70507602/146759085-ca4566f8-aa72-4f2e-b1f1-77bf2e7d2c7e.png)

**Comment by xgrau** [comment]
I have also encountered a similar issue with `samap` 0.3.3 — whenever I'm loading blast maps from two species or more, only the homologs from the first one are correctly parsed:

```python
dict_sam = { "sp1": sam1, "sp2": sam2, "sp3": sam3, "sp4": sam4 }
dict_clu = { "sp1": "leiden_clusters", "sp2": "leiden_clusters", "sp3": "leiden_clusters", "sp4": "leiden_clusters" }
samm_am = SAMAP(dict_sam, f_maps = "blastdbs/", keys = dict_clu)
```
Output:

```
Not updating the manifold...
Not updating the manifold...
Not updating the manifold...
Not updating the manifold...
12251 `sp1` gene symbols match between the datasets and the BLAST graph.
0 `sp2` gene symbols match between the datasets and the BLAST graph.
0 `sp3` gene symbols match between the datasets and the BLAST graph.
0 `sp4` gene symbols match between the datasets and the BLAST graph.
```

I've tracked down the issue to an incorrect parsing of the species labels that are preappended to gene names from the blast database when creating the `SAMAP` object: they are added to the first species during blast map parsing, but not to the subsequent ones. Then, when `SAMAP` attempts to match these gene names with those from the single-species `sam` objects, it returns zero shared gene ids.

I've manually checked whether my blast db is correctly formatted every which way, and couldn't find any issue with it: all genes from all four species have the correct prefix. 

However, I can't reproduce the issue with the example planarian and _Schistosoma_ data: in this case, it all works fine.

As a workaround, I've tweaked a workaround by tweaking the `prepend_var_prefix` function in `utils.py`:

```python
def prepend_var_prefix(s, pre):
    # x = [x.split("_")[0] for x in s.adata.var_names]
    # if not (np.unique(x).size == 1 and x[0] == pre):
    y = list(s.adata.var_names)
    vn = [pre + "_" + x for x in y]
    s.adata.var_names = vn
    s.adata_raw.var_names = vn
```

Basically, this forces the addition of species prefixes to the genes parsed from the blast maps. As far as I can tell this will only work if you manually ensure that gene names from the `sam` objects and the blast maps all have species prefixes right from the start.

By the way, I'm using the following software versions (from `conda`):

```bash
anndata                   0.7.8            py37h89c1867_1    conda-forge
sam-algorithm             0.8.9              pyhdfd78af_0    bioconda
samap                     0.3.3              pyhdfd78af_0    bioconda
```

Cheers

**Comment by atarashansky** [comment]
Thanks for catching this! I'll fix and update releases ASAP.

**Comment by atarashansky** [comment]
Would it be possible to send 2 h5ad's (just a few cells each) and the blast output so I can try and reproduce?

**Comment by atarashansky** [comment]
![image](https://user-images.githubusercontent.com/16548075/149804020-8676558b-1b91-44fa-a8fe-9e1529e471f0.png)
Your BLAST tables look like the above, correct?

**Comment by atarashansky** [comment]
Perhaps try installing from source in a new environment. If the error still persists, then I would need data to reproduce it before fixing. 

```
conda create -n samap_test python=3.7
conda activate samap_test 
git clone https://github.com/atarashansky/SAMap.git samap_directory
cd samap_directory
pip install .
pip install jupyterlab
```

**Comment by xgrau** [comment]
Hi Alexander, 

Thanks for your prompt answer! I've tried to create a small dataset where the problem can be reproduced — you can find it here: https://www.dropbox.com/sh/stn1h1vziteeh6p/AACdGxKCE0yxKmpYAMCF80Dua?dl=0

I've run the `test.py` script using the software versions pulled from your command above.

```
anndata                   0.7.8                    pypi_0    pypi
scanpy                    1.8.2                    pypi_0    pypi
sam-algorithm             0.8.9                    pypi_0    pypi
samap                     0.3.4                    pypi_0    pypi
```

> What OS are you running on?

I've tried it in Ubuntu 20.04 LTS and Scientific Linux 7.2 (Nitrogen), with an identical error (same software versions).

Hope you find this useful.

Cheers.

**Comment by atarashansky** [comment]
@xgrau 

Please install `samap==0.3.4` from PIP (or from github main branch) and let me know if your issue has been resolved.

Thanks for your patience,
Alec

**Comment by xgrau** [comment]
Hi Alec,

It's indeed been resolved, thanks a lot!

```bash
(samap_v0.3.4) [test-samap-xgb]: python test.py 
Not updating the manifold...
Not updating the manifold...
8084 `sp1` gene symbols match between the datasets and the BLAST graph.
8292 `sp2` gene symbols match between the datasets and the BLAST graph.
```

**Comment by huawen-poppy** [comment]
Hello Alexander, 
Thank you for your kindly follow-up on this issue. I tried to create a new environment and reinstalled everything to perform the analysis following your suggestions. But I still face the same problem. I created some test data for reproducing the problem. You can find them in [my dropbox](https://www.dropbox.com/sh/8da9w5mnwo9fijt/AABrSjEplWkC_sUqVKBNzTZqa?dl=0).
Thanks in advance!

**Comment by atarashansky** [comment]
Can you provide the BLAST output as well? (the maps/ directory)

**Comment by atarashansky** [comment]
@huawen-poppy 

It was a very niche bug, but I figured out what was causing your issue. The `adata.raw` attribute in your AnnData had `adata.raw.var_names` equal to integers instead of the gene IDs so the resulting `sam.adata_raw` AnnData generated had the wrong (integer) gene names.

In any case, I ensured that the gene names for the raw counts would be taken from `adata` instead of `adata.raw`. Please update SAM with:

`pip install samalg==0.9.0` and your problem should be resolved. Thanks for your patience!

**Comment by huawen-poppy** [comment]
Hello, Alex. Thank you for the quick reply. I have uploaded the BLAST output files. I tried to update SAM but failed with the errors. It seems like it is not released yet.

```
ERROR: Could not find a version that satisfies the requirement samalg==0.9.0 (from versions: none)
ERROR: No matching distribution found for samalg==0.9.0
```

**Comment by atarashansky** [comment]
Shoot, sorry, it’s 

pip install sam-algorithm==0.9.0

**Comment by huawen-poppy** [comment]
Hi, Alex. The problem has been solved! Thank you very much!! :)

**Comment by xiangyupan** [comment]
Hi @atarashansky 
Sorry to border you, I met the same error. The code is as below.
![图片](https://user-images.githubusercontent.com/20773767/236370373-567051da-f7a8-4dd9-800a-35c6f7461174.png)
Following your step by step test above, I also checked the version of my samap, which is the latest one (v1.0.0). Then other tests were also done as follow.
![图片](https://user-images.githubusercontent.com/20773767/236370484-b7550084-db48-4541-8c2c-14f2cea2bf80.png)
![图片](https://user-images.githubusercontent.com/20773767/236370546-652dbc21-1b16-41ab-b7a7-55cf19c554f7.png)
![图片](https://user-images.githubusercontent.com/20773767/236370570-3ede23eb-168c-4758-8ebc-01623910adc2.png)
I am not sure what's wrong with my data. Hope to your reply. Many thanks.
